core/arch/riscv/kernel/entry.S

```diff
@@ -0,0 +1,221 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2022-2023 NXP
+ */
+
+#include <asm.S>
+#include <generated/asm-defines.h>
+#include <keep.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <riscv.h>
+#include <riscv_macros.S>
+
+.section .data
+.balign 4
+
+#ifdef CFG_BOOT_SYNC_CPU
+.equ SEM_CPU_READY, 1
+#endif
+
+	/*
+	 * Setup sp to point to the top of the tmp stack for the current CPU:
+	 * sp is assigned:
+	 * stack_tmp + cpu_id * stack_tmp_stride - STACK_TMP_GUARD
+	 */
+.macro set_sp
+	/* Unsupported CPU, park it before it breaks something */
+	li	t1, CFG_TEE_CORE_NB_CORE
+	csrr	t0, CSR_XSCRATCH
+	bge	t0, t1, unhandled_cpu
+	la	t2, stack_tmp
+	lw	t1, stack_tmp_stride
+	/*
+	 * t0 is core pos + 1
+	 * t1 is value of stack_tmp_stride
+	 * t2 is value of stack_tmp
+	 */
+	mv	sp, t2
+	mul	t2, t1, t0
+	add	sp, sp, t2
+.endm
+
+.macro cpu_is_ready
+#ifdef CFG_BOOT_SYNC_CPU
+	csrr	t0, CSR_XSCRATCH
+	la	t1, sem_cpu_sync
+	slli	t0, t0, 2
+	add	t1, t1, t0
+	li	t2, SEM_CPU_READY
+	sw	t2, 0(t1)
+	fence
+#endif
+.endm
+
+.macro set_tp
+	csrr	a0, CSR_XSCRATCH
+	li	a1, THREAD_CORE_LOCAL_SIZE
+	la	tp, thread_core_local
+	mul	a2, a1, a0
+	add	tp, tp, a2
+	sw	a0, THREAD_CORE_LOCAL_HART_ID(tp)
+.endm
+
+.macro set_satp
+	la	a1, boot_mmu_config
+	LDR	a0, CORE_MMU_CONFIG_SATP(a1)
+	csrw	CSR_SATP, a0
+	sfence.vma	zero, zero
+.endm
+
+.macro wait_primary
+#ifdef CFG_BOOT_SYNC_CPU
+	la	t0, sem_cpu_sync
+	li	t2, SEM_CPU_READY
+1:
+	fence	w, w
+	lw	t1, 0(t0)
+	bne	t1, t2, 1b
+#endif
+.endm
+
+.macro wait_secondary
+#ifdef CFG_BOOT_SYNC_CPU
+	la	t0, sem_cpu_sync
+	li	t1, CFG_TEE_CORE_NB_CORE
+	li	t2, SEM_CPU_READY
+1:
+	addi	t1, t1, -1
+	beqz	t1, 3f
+	addi	t0, t0, 4
+2:
+	fence
+	lw	t1, 0(t0)
+	bne	t1, t2, 2b
+	j	1b
+3:
+#endif
+.endm
+
+#ifdef CFG_BOOT_SYNC_CPU
+#define flush_cpu_semaphores \
+		la	t0, sem_cpu_sync_start
+		la	t1, sem_cpu_sync_end
+		fence
+#else
+#define flush_cpu_semaphores
+#endif
+
+.macro bootargs_entry
+	/*
+	 * Save boot arguments
+	 */
+	la	t0, boot_args
+	/* Save boot hart */
+	STR	a0, REGOFF(0)(t0)
+	/* Save FDT address */
+	STR	a1, REGOFF(1)(t0)
+.endm
+
+FUNC _start , :
+.option push
+.option norelax
+	la	gp, __global_pointer$
+.option pop
+#ifdef CFG_RISCV_M_MODE
+	csrr	a0, CSR_MHARTID
+#endif
+	csrw	CSR_XSCRATCH, a0
+	bnez	a0, reset_secondary
+	jal	reset_primary
+	j	.
+END_FUNC _start
+
+LOCAL_FUNC reset_primary , : , .identity_map
+UNWIND(	.cantunwind)
+
+	bootargs_entry
+
+	/*
+	 * Zero bss
+	 */
+1:
+	lla	t0, __bss_start
+	lla	t1, __bss_end
+	beq	t0, t1, 1f
```

Comment:
shouldn't this be `1b` not `1f`?

Reply:
is `1:` missing at line 150?

Reply:
Yes, added and removed the top `1:` label.

Reply:
ok, right. latest post is fine.